runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -68,83 +71,107 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
+                    return map1;
+                });
+
+        return waitOrHoleFill(runtimeLayout, readResult, waitForWrite);
     }
 
-    /**
-     * Reads a range of addresses from the Chain of log unit servers. This method optimizes for the time to wait to
-     * hole fill in case an empty address is encountered.
-     * If the waitForWrite flag is set to true, when an empty address is encountered, it waits for the hole to be
-     * filled. All subsequent empty addresses are hole filled directly and the reader does not wait.
-     * In case the flag is set to false, none of the reads wait for write completion and the empty address is hole
-     * filled right away.
-     *
-     * @param runtimeLayout Runtime layout.
-     * @param range         Range of addresses to read.
-     * @param waitForWrite  Flag whether wait for write is required or hole fill directly.
-     * @return Map of read addresses.
-     */
-    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Range<Long> range, boolean waitForWrite) {
-        long startAddress = range.lowerEndpoint();
-        long endAddress = range.upperEndpoint();
-        int numUnits = runtimeLayout.getLayout().getSegmentLength(startAddress);
-        log.trace("readRange[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
+    private Map<Long, ILogData> waitOrHoleFill(RuntimeLayout runtimeLayout,
+                                               Map<Long, LogData> readResult,
+                                               boolean waitForWrite) {
+        Map<Long, ILogData> returnResult = new TreeMap<>(readResult);
 
-        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(
-                runtimeLayout
-                        .getLogUnitClient(startAddress, numUnits - 1)
-                        .read(range)).getAddresses();
+        // If waiting not required for writer, hole fill directly
+        if (!waitForWrite) {
+            readResult.forEach((address, value) -> {
+                if (value.isEmpty()) {
+                    holeFill(runtimeLayout, address);
+                    returnResult.put(address, peek(runtimeLayout, address));
+                }
+            });
+            return returnResult;
+        }
 
         // In case of holes, use the standard backoff policy for hole fill for
-        // the first entry in the list. All subsequent holes in the list can
-        // be hole filled without waiting as we have already waited for the first
-        // hole.
-        boolean wait = !waitForWrite;
-        Map<Long, ILogData> returnResult = new TreeMap<>();
-        for (Map.Entry<Long, LogData> entry : logResult.entrySet()) {
+        // the first entry in the list. All subsequent holes in the list can be
+        // hole filled without waiting as we have already waited for the first hole.
+        boolean wait = true;
+        for (Map.Entry<Long, LogData> entry : readResult.entrySet()) {
```

Comment:
Shouldn't this actually iterate on returnResult?
Please refactor this, changing status values inline is not clear.

Maybe we can do this:
1. Get largest address from the sorted read result
2. Apply read (with hole fill policy) on address from #1 
3. Iterate the sorted read result and try to fill hole from high to low

Also, instead of peek, then hole fill, you can hole fill directly and ignore OverWrite exceptions.

Reply:
As discussed, for range read, waiting on last address or an arbitrary one are conceptually and effectively same.